prometheus/promhttp/instrument_server.go

```diff
@@ -0,0 +1,442 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (c) 2013, The Prometheus Authors
+// All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found
+// in the LICENSE file.
+
+package promhttp
+
+import (
+	"bufio"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+// InFlight accepts a Gauge and an http.Handler, returning a new http.Handler
+// that wraps the supplied http.Handler. The provided Gauge must be registered
+// in a registry in order to be used.
+func InFlight(g prometheus.Gauge, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		g.Inc()
+		next.ServeHTTP(w, r)
+		g.Dec()
+	})
+}
+
+// Latency accepts an ObserverVec interface and an http.Handler, returning a
+// new http.Handler that wraps the supplied http.Handler. The provided
+// ObserverVec must be registered in a registry in order to be used.
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, the supplied ObserverVec will report a 200.  The instance
+// labels "code" and "method" are supported on the provided ObserverVec.
+// Note: Partitioning histograms is expensive.
+func Latency(obs prometheus.ObserverVec, next http.Handler) http.Handler {
+	code, method := checkLabels(obs)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var (
+			status int
+			now    = time.Now()
+		)
+		if code {
+			d := newDelegator(w)
+			next.ServeHTTP(d, r)
+			status = d.Status()
+		} else {
+			next.ServeHTTP(w, r)
+		}
+
+		obs.With(labels(code, method, r.Method, status)).Observe(time.Since(now).Seconds())
```

Comment:
Disregard, I understand your checkLabels workaround now.

I guess it could be an option to move the `if code` block outside of the Handlerfunc and return two different HandlerFuncs here. It probably won't have a huge impact, but it would save a few conditional checks during every request.